### PR TITLE
execsnoop: reliably fetch arguments and environ variable on arm64

### PR DESCRIPTION
**Pull request by lebauce**
Accessing argv and envp parameters of execve syscall can cause minor page faults
on arm64, which cause bpf_probe_user_read functions to return EFAULT. Therefore,
we hook into either 'prepare_binprm' or 'bprm_execve' functions that are called
just after the kernel accessed it first.

### CODE REVIEW

File: tools/execsnoop.py [hunk 1]
```diff
@@ -227,6 +259,10 @@ def parse_uid(user):
 b = BPF(text=bpf_text)
 execve_fnname = b.get_syscall_fnname("execve")
 b.attach_kprobe(event=execve_fnname, fn_name="syscall__execve")
+if BPF.get_kprobe_functions(b'prepare_binprm'):
+    b.attach_kprobe(event="prepare_binprm", fn_name="parse_argv_envp")
+else:
+    b.attach_kprobe(event="bprm_execve", fn_name="parse_argv_envp")
```
Comment (yonghong-song) [inline tools/execsnoop.py:265]: In latest kernel (I checked bpf-next only), both prepare_binprm and bprm_execve become static variable and prepare_binprm became a much smaller function and easily subject to inlining.

Since we are saving pointer and wait until a later pointer to do the actually probe_read. Can we do this during return probe?

Comment (lebauce) [reply tools/execsnoop.py:265]: When the `execve` succeeds, the return probe is never executed (the kretprobe seems to be there only to get the args of `execve` that failed).

`prepare_binrpm` is supposed to be used on old kernels where it was an exported symbol using EXPORT_SYMBOL (I missed the fact that `prepare_binrpm` is still present in early kernels but almost always inlined as you said because it's very small).

`bprm_execve` is indeed declared static but almost never inlined as it's pretty big. We could switch to using `prepare_exec_creds` which is declared as `extern` and called shortly after the args were accessed by the kernel.

To sum up, I suggest trying to first attach to `prepare_exec_creds` and, if it fails, attach to `prepare_binprm`. WDYT ?

Comment (yonghong-song) [reply tools/execsnoop.py:265]: > When the execve succeeds, the return probe is never executed (the kretprobe seems to be there only to get the args of execve that failed).

This seems not true for me on x64. I see
```
$ sudo trace.py 'r::__x64_sys_execve "retval=%d", retval'
PID     TID     COMM            FUNC             -
2987496 2987496 env             __x64_sys_execve retval=0
2987497 2987497 env             __x64_sys_execve retval=0
2987498 2987498 env             __x64_sys_execve retval=0
2987496 2987496 timeout         __x64_sys_execve retval=0
2987497 2987497 timeout         __x64_sys_execve retval=0
2987498 2987498 timeout         __x64_sys_execve retval=0
2987500 2987500 sh              __x64_sys_execve retval=0
2987499 2987499 sh              __x64_sys_execve retval=0
2987501 2987501 sh              __x64_sys_execve retval=0
2987499 2987499 wdb-up.sh       __x64_sys_execve retval=0
2987501 2987501 check_binary_ru __x64_sys_execve retval=0
2987502 2987502 sar             __x64_sys_execve retval=0
2987503 2987503 awk             __x64_sys_execve retval=0
```

The function latency indicates these short live processes takes
average 0.2 millisecond,
```
$ sudo funclatency.py __x64_sys_execve
Tracing 1 functions for "__x64_sys_execve"... Hit Ctrl-C to end.
^C
     nsecs               : count     distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 0        |                                        |
        32 -> 63         : 0        |                                        |
        64 -> 127        : 0        |                                        |
       128 -> 255        : 0        |                                        |
       256 -> 511        : 0        |                                        |
       512 -> 1023       : 0        |                                        |
      1024 -> 2047       : 2        |                                        |
      2048 -> 4095       : 6        |*                                       |
      4096 -> 8191       : 10       |*                                       |
      8192 -> 16383      : 0        |                                        |
     16384 -> 32767      : 0        |                                        |
     32768 -> 65535      : 0        |                                        |
     65536 -> 131071     : 56       |**********                              |
    131072 -> 262143     : 215      |****************************************|
    262144 -> 524287     : 17       |***                                     |
    524288 -> 1048575    : 1        |                                        |
   1048576 -> 2097151    : 0        |                                        |
   2097152 -> 4194303    : 7        |*                                       |

avg = 205112 nsecs, total: 64610445 nsecs, count: 315
```

so arm64 behave differently?

> To sum up, I suggest trying to first attach to prepare_exec_creds and, if it fails, attach to prepare_binprm. WDYT ?

`prepare_exec_creds` indeed looks better. But I would like to get a confirmation execve retprobe indeed not working in arm64 compared to x86_64.
